### PR TITLE
test: macros checks of attribute macros usages

### LINKED ISSUE #250: macros: Extend checks of attribute macros usages
### File Location(s)

_No response_

### Proposal

Cover edge cases

1. multiple `gservice` on impl
```rust
#[gservice]
#[gservice]
impl MyService {}
```

```rust
#[gservice]
impl MyService {
    pub fn func_1() {}
}

#[gservice]
impl MyService {
    pub fn func_2() {}
}
```

2. multiple `gprogram` on impl

```rust
#[gprogram]
#[gprogram]
impl MyProgram {
    pub fn new() -> Self {
        Self
    }
}
```

```rust
#[gprogram]
impl MyProgram {
    pub fn new() -> Self {
        Self
    }
}

#[gprogram]
impl MyProgram {
    pub fn default() -> Self {
        Self
    }
}
```

### PR DESCRIPTION
**Pull request by vobradovich**
Resolves #250

### CODE REVIEW

File: macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr [hunk 1]
```diff
@@ -0,0 +1,14 @@
+error: multiple `gservice` attributes are not allowed
+ --> tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.rs:7:1
+  |
+7 | impl MyService {
+  | ^^^^
+
+error[E0599]: no method named `this` found for struct `MyService` in the current scope
```
Comment (NikVolf) [inline macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr:7]: is it the actual error message that user will see?

if yes, can we remove further errors because macros is invalid on the first place?

Comment (DennisInSky) [reply macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr:7]: Wondered as well why this message pops out there

Comment (vobradovich) [reply macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr:7]: that's the second error, trying to figure how to remove it

Comment (DennisInSky) [reply macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr:7]: Weird, it should stop as soon as `abort!` is called. Would be nice if we can get rid of this second message

File: macros/core/src/program.rs [hunk 2]
```diff
@@ -8,19 +8,42 @@ use proc_macro_error::abort;
...
+    Type, TypePath, Visibility,
 };
 
+pub fn gprogram_safe(program_impl_tokens: TokenStream2) -> TokenStream2 {
+    let program_impl = parse_program_impl(program_impl_tokens);
+    check_program_single(&program_impl);
+    gen_gprogram_impl(program_impl)
+}
+
 pub fn gprogram(program_impl_tokens: TokenStream2) -> TokenStream2 {
```
[the remark this answers is not in the document]
Comment (holykol) [reply macros/core/src/program.rs:21]: Also consider `#[doc(hidden)]`

[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/program.rs]: fixed

File: macros/core/src/program.rs [hunk 3]
```diff
@@ -8,19 +8,42 @@ use proc_macro_error::abort;
...
         abort!(
             err.span(),
             "`gprogram` attribute can be applied to impls only: {}",
             err
         )
-    });
+    })
+}
+
+fn check_program_single(program_impl: &ItemImpl) {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/program.rs]: fixed

File: macros/core/src/program.rs [hunk 4]
```diff
@@ -8,19 +8,42 @@ use proc_macro_error::abort;
...
+    gen_gprogram_impl(program_impl)
+}
+
 pub fn gprogram(program_impl_tokens: TokenStream2) -> TokenStream2 {
-    let program_impl = syn::parse2(program_impl_tokens).unwrap_or_else(|err| {
+    let program_impl = parse_program_impl(program_impl_tokens);
+    gen_gprogram_impl(program_impl)
+}
+
+fn parse_program_impl(program_impl_tokens: TokenStream2) -> ItemImpl {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/program.rs]: fixed

File: macros/core/src/shared.rs [hunk 5]
```diff
@@ -8,6 +8,11 @@ use syn::{
     ReturnType, Signature, Type, TypePath, TypeTuple, WhereClause,
 };
 
+pub(crate) static mut SERVICE_TYPES: BTreeMap<String, Span> = BTreeMap::new();
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/shared.rs]: moved to service.rs

File: macros/core/src/shared.rs [hunk 6]
```diff
@@ -8,6 +8,11 @@ use syn::{
...
+
+/// Static Span of Program `impl` block
+pub(crate) static mut PROGRAM_SPAN: Option<Span> = None;
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/shared.rs]: moved to program.rs

File: macros/core/src/service.rs [hunk 7]
```diff
@@ -29,19 +29,44 @@ use proc_macro_error::abort;
 use quote::quote;
 use std::collections::BTreeMap;
 use syn::{
-    GenericArgument, Ident, ImplItemFn, ItemImpl, Path, PathArguments, Type, TypeParamBound,
-    Visibility, WhereClause, WherePredicate,
+    spanned::Spanned, GenericArgument, Ident, ImplItemFn, ItemImpl, Path, PathArguments, Type,
+    TypeParamBound, Visibility, WhereClause, WherePredicate,
 };
 
+pub fn gservice_safe(service_impl_tokens: TokenStream2) -> TokenStream2 {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/service.rs]: fixed

File: macros/core/src/service.rs [hunk 8]
```diff
@@ -29,19 +29,44 @@ use proc_macro_error::abort;
...
+    gen_gservice_impl(service_impl)
+}
+
 pub fn gservice(service_impl_tokens: TokenStream2) -> TokenStream2 {
-    let service_impl = syn::parse2(service_impl_tokens).unwrap_or_else(|err| {
+    let service_impl = parse_service_impl(service_impl_tokens);
+    gen_gservice_impl(service_impl)
+}
+
+fn parse_service_impl(service_impl_tokens: TokenStream2) -> ItemImpl {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/service.rs]: fixed

File: macros/core/src/service.rs [hunk 9]
```diff
@@ -29,19 +29,44 @@ use proc_macro_error::abort;
...
         abort!(
             err.span(),
             "`gservice` attribute can be applied to impls only: {}",
             err
         )
-    });
+    })
+}
+
+fn check_service_single_by_name(service_impl: &ItemImpl) {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/core/src/service.rs]: fixed

File: macros/tests/ui/gservice_fails_multiple_not_allowed.stderr [hunk 10]
```diff
@@ -0,0 +1,5 @@
+error: multiple `gservice` attributes are not allowed
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply macros/tests/ui/gservice_fails_multiple_not_allowed.stderr]: fixed

File: macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr (hunk 1 above)
Comment (vobradovich) [reply macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr:7]: I think the first attribute has already generated some code, then the second call `abort!`

Comment (vobradovich) [reply macros/tests/ui/gservice_fails_multiple_not_allowed_on_one_impl.stderr]: added new method  `ensure_single_gservice_on_impl`

**Review by DennisInSky — APPROVED** [review]
